src/librustc_typeck/check/cast.rs

```diff
@@ -139,6 +140,21 @@ impl<'a, 'gcx, 'tcx> CastCheck<'tcx> {
 
     fn report_cast_error(&self, fcx: &FnCtxt<'a, 'gcx, 'tcx>, e: CastError) {
         match e {
+            CastError::NeedDeref => {
+                let cast_ty = fcx.ty_to_string(self.cast_ty);
+                let mut err = fcx.type_error_struct(self.cast_span,
```

Comment:
Pointing at the `self.cast_span` instead of `self.span` (`i16` instead of `s as i16` in the example) because when trying to apply the secondary label on just `self.expr.span` (`s` in the example) the actual output looks like this:

```nocode
error: casting `&f64` as `i16` is invalid
 --> file3.rs:2:30
  |
2 |     vec![0.0].iter().map(|s| s as i16).collect::<Vec<i16>>();
  |                              ^^^^^^^^
  |                              |
  |                              casting `&f64` as `i16` is invalid      // <<< this in red
  |                              try dereferencing for the cast to work  // <<< this in blue
```

losing the fact that the secondary label's hint should only apply to the `s` only.